docs/root/api-v1/http_filters/ip_tagging_filter.rst

```diff
@@ -0,0 +1,63 @@
+.. _config_http_filters_ip_tagging_v1:
+
+HTTP IP Tagging Filter
+======================
+
+HTTP IP Tagging :ref:`configuration overview <config_http_filters_ip_tagging>`.
+
+.. code-block:: json
+
+  {
+    "name": "ip_tagging",
+    "config": {
+      "request_type": "...",
+      "fill_factor" : "...",
+      "root_branching_factor": "...",
+      "ip_tags": []
+    }
+  }
+
+request_type
+  *(optional, string)* The type of requests the filter should apply to. The supported
+  types are *INTERNAL*, *EXTERNAL* or *BOTH*. A request is considered internal if
+  :ref:`x-envoy-internal<config_http_conn_man_headers_x-envoy-internal>` is set to true. If
+  :ref:`x-envoy-internal<config_http_conn_man_headers_x-envoy-internal>` is not set or false, a
+  request is considered external. The filter defaults to *both*, and it will apply to all request
+  types. Default value is *BOTH*.
+
+fill_factor
+  *(optional, double)* Supplies the fraction of completeness to use when calculating the branch. The supported range is
```

Comment:
Curious on your thoughts on whether this belongs as an API delivered configuration, hardcoded into the C++ or on the runtime filesystem? This seems an option for advanced users who want to tune the tree structure in close conjunction with the tag ranges, is there a concrete motivation here at Lyft?

Reply:
It is definitely an advanced feature; however I do think if larger data sets are used this would be good to modify. The benefits of modifying these improve performance. I can call that out more explicitly.  If they are specified, these values are configurable I think they should be as close as possible to the set of IP tags in the API configuration.

If we want to stick with defaults for now, I would suggest 0.5 for fill factor and 16 for root branching factor (based on the values referenced in the LC Trie paper).

@brian-pane Do you have any use cases for modifying the fill factor and root branching factor?

Reply:
I think it would be reasonable to just hard-code a fill factor of 0.5 in the C++ code. Here's my rationale: In the absolute worst case, a fill factor of 0.5 will result in the trie using an extra 1MB of memory (2^19 nodes, each 4 bytes in size, and half of the nodes only exist because of the fill factor of 0.5) to reduce the tree height.

Reply:
If root branching factor is a count of bits (rather than a count of child nodes), the original paper's recommendation is arguably too big for common Envoy usage. The authors' use case was a routing table containing tens or hundreds of thousands of prefixes. For a proxy, in contrast, using 4 or 8 bits for the initial branching seems sufficient.

Reply:
Okay, i'll remove fill factor and root branching factor and hard code the values to 0.5  and 16 for respectively. I won't change the interface of LC trie incase in the future we want them to be customizable.